librepo/gpg_rpm.c

```diff
@@ -0,0 +1,613 @@
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <rpm/rpmpgp.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "gpg.h"
+#include "gpg_internal.h"
+#include "rcodes.h"
+#include "util.h"
+
+static const char * const BEGIN_OPENPGP_BLOCK = "-----BEGIN PGP ";
+static const char * const BEGIN_OPENPGP_PUBKEY_BLOCK = "-----BEGIN PGP PUBLIC KEY BLOCK-----";
+static const char * const BEGIN_OPENPGP_SIGNATURE = "-----BEGIN PGP SIGNATURE-----";
+
+static const char *
+search_line_starts_with(const char *haystack, size_t len, const char *needle) {
+    const size_t needle_len = strlen(needle);
+    const char *line = haystack;
+    size_t len_to_end = len;
+    while (len_to_end >= needle_len) {
+        if (strncmp(line, needle, needle_len) == 0) {
+            return line;
+        }
+        line = memchr(line, '\n', len_to_end);
+        if (line == NULL) {
+            break;
+        }
+        ++line;
+        len_to_end = len - (line - haystack);
+    }
+    return NULL;
+}
+
+static void
+ensure_dir_exists(const char * dir)
+{
+    const int res = mkdir(dir, 0744);
+    if (res != 0 && errno != EEXIST) {
+        g_warning("Failed to create \"%s\": %d - %s\n", dir, errno, g_strerror(errno));
+    }
+}
+
+static inline gchar
+nibble_to_hex(guint8 value) {
+    return value < 10 ? value + '0' : value - 10 + 'A';
+}
+
+static gchar *
+bin_to_hex(const guint8 * buffer, gsize buffer_length) {
+    gchar * const ret = g_new(gchar, buffer_length * 2 + 1);
+    gchar * out_ptr = ret;
+    for (gsize i = 0; i < buffer_length; ++i) {
+        guint8 value = buffer[i];
+        *out_ptr++ = nibble_to_hex(value >> 4);
+        *out_ptr++ = nibble_to_hex(value & 0x0F);
+    }
+    *out_ptr = '\0';
+    return ret;
+}
+
+static ssize_t
+read_file_fd_to_memory(int fd, gchar ** buf, GError **err) {
+    const off_t size = lseek(fd, 0, SEEK_END);
+    if (size == -1) {
+        const gchar * errstr = g_strerror(errno);
+        g_debug("%s: Can't seek to end of file: %s", __func__, errstr);
+        g_set_error(err, LR_GPG_ERROR, LRE_IO, "Can't seek to end of file: %s", errstr);
+        return -1;
+    }
+    if (lseek(fd, 0, SEEK_SET) == -1) {
+        const gchar * errstr = g_strerror(errno);
+        g_debug("%s: Can't seek to beginning of file: %s", __func__, errstr);
+        g_set_error(err, LR_GPG_ERROR, LRE_IO, "Can't seek to beginning of file: %s", errstr);
+        return -1;
+    }
+    g_autofree gchar * bufp = g_new(gchar, size);
+    ssize_t ret = read(fd, bufp, size);
+    if (ret == -1) {
+        const gchar * errstr = g_strerror(errno);
+        g_debug("%s: Error reading from file descriptor %i: %s", __func__, fd, errstr);
+        g_set_error(err, LR_GPG_ERROR, LRE_IO, "Error reading from file descriptor %i: %s", fd, errstr);
+        return -1;
+    }
+    if (ret != size) {
+        g_debug("%s: Detected file size %li but readed %li", __func__, (long)size, (long)ret);
```

Comment:
Typo: `readed` -> `but %li was read`. Same in the line below...

Reply:
Thanks